Diff updates

mrollinsiv opened the pull request:
- Added MIT License (Fixes #1)
- Literal Array test
- Flags for ignoreKey & falsy
- JSDocs and package information

mrollinsiv committed:
Updated README.md to detail new options parameter

mrollinsiv committed:
Logic to handle 3-way-diff of array of literal values

mrollinsiv committed:
Added .idea directory to .gitignore

mrollinsiv committed:
3-way-diff for arrays of literals
Created new tests for arrays.  Renamed index.js to literals.js based on new array tests.

mrollinsiv committed:
Fixed undefined options in recursive call

mrollinsiv committed:
ignoreKey flag
Add logic to handle ignoring specific keys when running diff. Updated docs.

mrollinsiv committed:
JSDocs, package info
Added full JSDoc notation, created authors in package and module.

mrollinsiv committed:
Falsy test
Allow path to be flag as falsy to allow comparison where any false value is equivalent.  Updated docs and tests.

mrollinsiv committed:
Removed 'A' return `kind`

mrollinsiv committed:
Added MIT License

shawnpwalsh commented on README.md:
```diff
@@ -21,11 +21,49 @@ Differences are reported as one or more change records. Change records have the
+* `item` - when kind === 'A', contains a nested change record indicating the change that occurred at the array index
+
+### Options
+Options object can be used to denote certain flags.  Object should be replica of `mine` with flags set on specific keys
```
Not necessarily a complete replica, right?

shawnpwalsh commented on index.js:
```diff
@@ -1,6 +1,30 @@
+ *
+ * @param {object} parent - Object with parent values
+ * @param {object} theirs - Object with their changes
+ * @param {object} mine - Object with my changes
```
Both of these lines, we should probably use "values" instead of "changes"

shawnpwalsh commented on index.js:
```diff
@@ -1,6 +1,30 @@
+ * @returns {Array} - Array of objects specifying the differences
+ */
+module.exports = function(parent, theirs, mine, options) {
+  options = options || {};
```
As I suggested to @billmalarky (https://github.com/WennerMedia/fastly-promise/pull/2#discussion_r92510632) we should probably use ES6 default parameters instead of this line.

shawnpwalsh commented on index.js:
```diff
@@ -15,46 +39,135 @@ module.exports = function(parent, theirs, mine) {
+ * @returns {Array} - Array of objects specifying the differences
+ */
+function recurse(parent, theirs, mine, path, options) {
   var path = path || [];
```
Suggest using ES6 default parameters for path and options

shawnpwalsh commented on index.js:
```diff
@@ -15,46 +39,135 @@ module.exports = function(parent, theirs, mine) {
+      results = results.concat(processKeyValuePair(key, value, parent, theirs, mine, path, options));
+    }
+  });
+
```
What about keys that exist in theirs but not parent/mine?

mrollinsiv replied:
This was added in new lines that follow

shawnpwalsh replied:
But no new tests were written?

mrollinsiv replied:
Added tests

shawnpwalsh commented on index.js:
```diff
@@ -15,46 +39,135 @@ module.exports = function(parent, theirs, mine) {
+
+  // Only process keys that have no options, or have not been flagged as ignored
+  if (_.isUndefined(options[key]) || (!_.isUndefined(options[key]) && !options[key]['ignoreKey'])) {
+    if (!_.isArray(value) && !_.isObject(value)) {
```
Maybe we should reorder these clauses and make this an `else` just to simplify?

shawnpwalsh commented on index.js:
```diff
@@ -15,46 +39,135 @@ module.exports = function(parent, theirs, mine) {
+      }
+      _.forOwn(array, function (value, key) {
+        if (!_.isArray(value) && !_.isObject(value)) {
+
```
Probably need a TODO here?

shawnpwalsh commented on index.js:
```diff
@@ -15,46 +39,135 @@ module.exports = function(parent, theirs, mine) {
+      // If single dimension array run the comparison
+      if (value.filter(function (i) {
+          return !_.isArray(i) && !_.isObject(i);
+        }).length == value.length) {
```
This is a pretty complex `if` statement. Should probably pull this out to a variable

shawnpwalsh commented on index.js:
```diff
@@ -15,46 +39,135 @@ module.exports = function(parent, theirs, mine) {
+          return !_.isArray(i) && !_.isObject(i);
+        }).length == value.length) {
+        path.push(key);
+        var differences = compareValues(parent[key], theirs[key], mine[key], path, options[key] || {});
```
By setting ES6 default parameters for `compareValues` we can get rid of the awkward `|| {}` in this call.

shawnpwalsh commented on index.js:
```diff
@@ -15,46 +39,135 @@ module.exports = function(parent, theirs, mine) {
+ * @param {object} mine - Object with my changes
+ * @param {array} path - Path to value in parent/theirs/mine
+ * @param {object} options - Key specific options for diff
+ * @returns {*} - An object containing the conflict/edit, otherwise empty return
```
Never actually seen `{*}` before, is that actually the standard? Versus something like `{} || null`

fedgrant commented on index.js:
```diff
@@ -15,46 +39,135 @@ module.exports = function(parent, theirs, mine) {
+        if (differences) {
+          results = results.concat(differences);
+        }
+        path.pop(key);
```
why push in the key to pop it out?

mrollinsiv replied:
This occurs in a recursive function and passes "path" by reference so it continues to track where in the path the code is.

shawnpwalsh commented on index.js:
```diff
@@ -71,6 +184,46 @@ function compareValues(parent, theirs, mine, path) {
+    if (!_.isUndefined(options) && options['ignoreOrder']) {
+      parent = parent.sort();
+      mine = mine.sort();
+      theirs = theirs.sort();
```
They all need to be arrays in order to call `sort` on all of them, right? It's only checking if one of them is.

shawnpwalsh commented on index.js:
```diff
@@ -71,6 +184,46 @@ function compareValues(parent, theirs, mine, path) {
+        mine: mine
+      };
+    } else {
+      // All 3 are equal
```
`return;`? (Also in the other place in this function with `// All 3 are equal` (line 258 currently)

mrollinsiv replied:
These were all added

mrollinsiv committed:
Removed the remaining 'A' kind documentation as it is not being used.

shawnpwalsh commented on test/arrays.js:
```diff
@@ -0,0 +1,117 @@
+var diff = require('../index.js');
+
+describe('Array Diff', function() {
+  it('both children edit same key to same array values', function() {
```
The children aren't "edit"ing anything?

mrollinsiv replied:
Correct, similar to "both children edit same key to same values" you wrote in literals.js.  Makes sure that valid change is passed.

shawnpwalsh replied:
Sorry, I meant that the children aren't editing anything in the code of this test (if they were, I'd be expecting them to be different from the parent)

mrollinsiv committed:
Babel
Installed babel to test against ES6

mrollinsiv committed:
Test array to literal changes

mrollinsiv committed:
PR Feedback
* Iterate over their values
* ES6 default parameter values
* Check for _.isArray before sort
* Add returns to no conflict/edit blocks
* Change order of value checks in processKeyValuePair to have old first check as the default else

mrollinsiv committed:
Merge branch 'master' into diff-updates

mrollinsiv committed:
Added .babelrc file

mrollinsiv committed:
Merge branch 'master' into diff-updates

mrollinsiv committed:
Cache node_modules for travis-ci

shawnpwalsh commented on index.js:
```diff
@@ -15,46 +37,140 @@ module.exports = function(parent, theirs, mine) {
+ * @param {object} options - Key specific options for diff
+ * @returns {Array} - Array of objects specifying the differences
+ */
+function recurse(parent, theirs, mine, path = [], options) {
```
No default param value for `options` here?

mrollinsiv committed:
Move slack to travis

mrollinsiv committed:
Fix env variable reference

mrollinsiv committed:
Remove secure

mrollinsiv committed:
Remove $ in env var

mrollinsiv committed:
Reverse cache and notifications

mrollinsiv committed:
Added default for options

mrollinsiv committed:
Add rooms to slack

mrollinsiv committed:
Updated tests
Added test for key in theirs, missing in parent/mine
Edited description of first array test

mrollinsiv committed:
Add $ to environment variables

mrollinsiv committed:
Added {} to env variable

mrollinsiv committed:
Revert to $SLACK_SECURE

mrollinsiv committed:
Removed Slack Notifications

shawnpwalsh commented on README.md:
`````diff
@@ -21,11 +21,47 @@ Differences are reported as one or more change records. Change records have the
+````
+
+Flags
+* `ignoreKey` - Do not diff this key
`````
It won't let me leave a comment on that line, but Line 12 TODO should be removed now that this is done

mrollinsiv committed:
Removed TODO for ignoring fields

mrollinsiv committed:
Merge remote-tracking branch 'origin/diff-updates' into diff-updates
